### PR TITLE
gh-118423: Add `INSTRUCTION_SIZE` macro to code generator

### PR DESCRIPTION
**Pull request by tomasr8**
I am not that familiar with the tier 1 interpreter (yet :smile: ) so I'd appreciate if someone could double check that for all
the places that use the new `INSTRUCTION_SIZE` macro, it is actually safe to do so.

I replaced all examples of `next_instr - this_instr`. There are still a few examples of `1 + INLINE_CACHE_ENTRIES_...` but I wasn't sure whether those were safe to change as well.

I named the new macro `INSTRUCTION_SIZE`, but I'm happy to change it to `INSTRUCTION_LENGTH` if you prefer.

Feedback welcome :)

* Issue: gh-118423

### CODE REVIEW

**Comment by tomasr8** [comment]
cc @iritkatriel since you asked for this in https://github.com/python/cpython/pull/118380#discussion_r1584469712 :)

File: Tools/cases_generator/generators_common.py [hunk 1]
```diff
@@ -490,12 +491,19 @@ def _emit_block(
...
                     self.out.emit(rbrace)
+                elif tkn.kind == "INSTRUCTION_SIZE":
+                    self._emit_instruction_size(tkn, inst)
                 else:
                     self.out.emit(tkn)
         except StackError as ex:
             raise analysis_error(ex.args[0], tkn) from None
         raise analysis_error("Expecting closing brace. Reached end of file", tkn)
 
+    def _emit_instruction_size(self, tkn: Token, inst: Instruction | None) -> None:
```
Comment (Fidget-Spinner) [inline Tools/cases_generator/generators_common.py]: This should be a `ReplacementFunctionType` and registered in the emitter as a replacement function. See for example `DECREF_INPUTS`

**Review by Fidget-Spinner — COMMENTED** [review]
This is really close to what I was thinking of. Great work!

File: Tools/cases_generator/lexer.py [hunk 2]
```diff
@@ -228,6 +228,9 @@ def choice(*opts: str) -> str:
     "tier2",
 }
 
+# An instruction size in the DSL
+INSTRUCTION_SIZE = "INSTRUCTION_SIZE"
+
```
Comment (Fidget-Spinner) [inline Tools/cases_generator/lexer.py]: Likewise, this doesn't have to be here.

File: Tools/cases_generator/generators_common.py [hunk 3]
```diff
@@ -357,6 +359,19 @@ def reload_stack(
...
 
+    def instruction_size(self,
+        tkn: Token,
+        tkn_iter: TokenIterator,
+        uop: Uop,
+        storage: Storage,
+        inst: Instruction | None
+    ) -> bool:
+        """Replace the INSTRUCTION_SIZE macro with the size of the current instruction."""
+        size = inst.size if inst else 1 + uop.size
```
Comment (tomasr8) [inline Tools/cases_generator/generators_common.py]: tier2 does not have access to the instruction here so I calculate the size as `1 + uop.size`. Does that make sense?

File: Tools/cases_generator/generators_common.py (hunk 1 above)
Comment (tomasr8) [reply Tools/cases_generator/generators_common.py]: Thanks for the advice! I modified the code to use the existing replacer logic

File: Tools/cases_generator/generators_common.py [hunk 4]
```diff
@@ -357,6 +359,18 @@ def reload_stack(
...
+        self.out.emit(tkn.replaceText(str(size)))
```
Comment (Fidget-Spinner) [inline Tools/cases_generator/generators_common.py]: Does `self.out.emit(f"{size}")` directly work?

Comment (tomasr8) [reply Tools/cases_generator/generators_common.py]: It does, but it does not insert white space around the instruction size, so you end up with

```int x =2;```
instead of 
```int x = 2;```

If that's acceptable, I can emit the size directly

Comment (Fidget-Spinner) [reply Tools/cases_generator/generators_common.py]: How about `self.out.emit(f" {size}")` then? Ideally, we want to keep this simple. `replaceText` confuses me a little.

Comment (tomasr8) [reply Tools/cases_generator/generators_common.py]: The thing is sometimes the space should be included and sometimes not, like inside parentheses:
```c
(INSTRUCTION_SIZE + oparg) -> (2 + oparg)
```
Same goes for whitespace that comes after `INSTRUCTION_SIZE`. We could simply do `self.out.emit(f" {size} ")` and call it a day, though we'll end up with some weird spacing in the generated code..

Comment (Fidget-Spinner) [reply Tools/cases_generator/generators_common.py]: Yup okay the latter sounds good. I rather a small bit of weirdness in the generated code than the generator being hard to understand. Thanks!

Comment (tomasr8) [reply Tools/cases_generator/generators_common.py]: updated :)

File: Tools/cases_generator/generators_common.py (hunk 3 above)
Comment (Fidget-Spinner) [reply Tools/cases_generator/generators_common.py]: Sorry this seems wrong, you need to emit the 1 + inline cache size not uop size

Comment (tomasr8) [reply Tools/cases_generator/generators_common.py]: Indeed, I see when comparing the output of `generated_cases.c.h` with `executor_cases.c.h` that the latter has the wrong instruction size. I fixed it (I think) by looking up the instructions for each `uop` in tier2 and passing it to the emitter.
Let me know if it's correct like that :)

**Comment by tomasr8** [comment]
It looks like there are not tests for tier2 yet so I only added tests for tier1

File: Tools/cases_generator/generators_common.py (hunk 3 above)
Comment (markshannon) [reply Tools/cases_generator/generators_common.py]: `INSTRUCTION_SIZE` is the size of the tier 1 instruction. This must be true even in tier 2.

> I fixed it (I think) by looking up the instructions for each uop in tier2 and passing it to the emitter.

That is correct, but it would be nice if we also checked that all matching tier 1 instructions are the same size.

File: Lib/test/test_generated_cases.py [hunk 5]
```diff
@@ -1371,6 +1371,24 @@ def test_stack_save_only(self):
...
+        TARGET(OP) {
+            frame->instr_ptr = next_instr;
+            next_instr += 1;
+            INSTRUCTION_STATS(OP);
+            frame->return_offset = 1 ;
+            DISPATCH();
+        }
+        """
+        self.run_cases_test(input, output)
+
```
Comment (markshannon) [inline Lib/test/test_generated_cases.py:1391]: Add a test for another instruction that is a different length, but uses `OP`:
```
macro(OP2) = unused/1 + OP;
```

File: Tools/cases_generator/generators_common.py [hunk 6]
```diff
@@ -1,3 +1,4 @@
+import dataclasses
```
Comment (markshannon) [inline Tools/cases_generator/generators_common.py]: I don't think this is used

File: Tools/cases_generator/tier2_generator.py [hunk 7]
```diff
@@ -167,7 +167,7 @@ def oparg(
         return True
 
 
-def write_uop(uop: Uop, emitter: Emitter, stack: Stack) -> Stack:
+def write_uop(uop: Uop, emitter: Emitter, stack: Stack, inst: Instruction | None) -> Stack:
```
Comment (markshannon) [inline Tools/cases_generator/tier2_generator.py]: Maybe pass the size, not the instruction. A uop can occur in many instructions, so picking one is arbitrary.
However, if it uses `INSTRUCTION_SIZE`, than all instructions must be the same size.

**Review by markshannon — CHANGES_REQUESTED** [review]
Thanks for doing this.

The general approach looks good, but I think we need to check that all the instructions containing the uop with `INSTRUCTION_LENGTH` are the same length.

File: Lib/test/test_generated_cases.py (hunk 5 above)
Comment (tomasr8) [reply Lib/test/test_generated_cases.py:1391]: Added!

File: Tools/cases_generator/tier2_generator.py (hunk 7 above)
Comment (tomasr8) [reply Tools/cases_generator/tier2_generator.py]: I added the check for the instruction size, which actually pointed out one place where we cannot use `INSTRUCTION_SIZE` instead of `next_instr - this_instr`:

https://github.com/python/cpython/blob/d8c864816121547338efa43c56e3f75ead98a924/Python/bytecodes.c#L4799

**Comment by tomasr8** [comment]
Seems like the CI is passing so if you could have another look :)

I have made the requested changes; please review again

File: Lib/test/test_generated_cases.py (hunk 5 above)
Comment (markshannon) [reply Lib/test/test_generated_cases.py:1391]: That test should fail, as the `INSTRUCTION_SIZE` is inconsistent.
The tier 2 `INSTRUCTION_SIZE` cannot be both 1 and 2.

**Comment by markshannon** [comment]
There seems to be some spurious spaces in the generated output, both before and after the number.

**Comment by tomasr8** [comment]
> There seems to be some spurious spaces in the generated output, both before and after the number.

Indeed, see this discussion: https://github.com/python/cpython/pull/125467#discussion_r1801262048
Ken preferred to keep the code simpler so it just emits a raw string which does not check for any existing white space. Though, it's also possible to wrap it in a `Token` instance which would emit the whitespace correctly, up to you :)

File: Lib/test/test_generated_cases.py (hunk 5 above)
Comment (tomasr8) [reply Lib/test/test_generated_cases.py:1391]: Right, sorry I mistakenly thought it should only fail for tier 2 hence I was only checking the instruction in the tier 2 generator. I extended the check to tier 1 as well and the test now fails as expected.

File: Tools/cases_generator/generators_common.py [hunk 8]
```diff
@@ -552,3 +583,35 @@ def cflags(p: Properties) -> str:
...
+    for inst in instructions.values():
+        if uop in inst.parts:
+            if size is None:
+                size = inst.size
+            if size != inst.size:
+                assert size == inst.size, (
+                    "All instructions containing a uop with the `INSTRUCTION_SIZE` macro "
+                    f"must have the same size: {size} != {inst.size}"
+                )
+    assert size is not None
```
Comment (markshannon) [inline Tools/cases_generator/generators_common.py]: This is not an fault in the code generator, but in the input. So it should raise and error, not be an assert.

**Comment by markshannon** [comment]
Apart for the one `assert` that should be a `raise`, this looks correct now.

Passing the instruction size or None to every replacement function seems a bit clunky though.
Maybe would could add an `instruction_size` attribute to the `Uop` class and do the instruction size checking in analyzer.py?

File: Tools/cases_generator/generators_common.py (hunk 8 above)
Comment (tomasr8) [reply Tools/cases_generator/generators_common.py]: Swapped for `raise analysis_error(...)` instead

**Comment by tomasr8** [comment]
> Passing the instruction size or None to every replacement function seems a bit clunky though.
Maybe would could add an instruction_size attribute to the Uop class and do the instruction size checking in analyzer.py?

I also thought it was less than ideal the way I did it, but I was a bit wary about adding more attributes to the classes. Though, your suggestion simplifies the code quite a bit!

I added a new attribute `Uop.instruction_size` which is only set for uops which contain the macro, otherwise it is `None`. This attribute is then used by the emitter.

**Comment by tomasr8** [comment]
Friendly ping @markshannon :slightly_smiling_face:  I think I addressed all of your points.

File: Lib/test/test_generated_cases.py [hunk 9]
```diff
@@ -1371,6 +1371,54 @@ def test_stack_save_only(self):
...
+        # Two instructions of different sizes referencing the same
+        # uop containing the `INSTRUCTION_SIZE` macro is not allowed.
+        input = """
+        inst(OP, (--)) {
+            frame->return_offset = INSTRUCTION_SIZE;
+        }
+        macro(OP2) = unused/1 + OP;
+        """
+
+        output = """
```
Comment (markshannon) [inline Lib/test/test_generated_cases.py]: There is no output. `output=""` should do.

File: Lib/test/test_generated_cases.py [hunk 10]
```diff
@@ -1371,6 +1371,54 @@ def test_stack_save_only(self):
...
+        TARGET(OP2) {
+            frame->instr_ptr = next_instr;
+            next_instr += 2;
+            INSTRUCTION_STATS(OP2);
+            /* Skip 1 cache entry */
+            frame->return_offset = 2 ;
+            DISPATCH();
+        }
+        """
+        with self.assertRaisesRegex(AssertionError, "All instructions containing a uop"):
```
Comment (markshannon) [inline Lib/test/test_generated_cases.py]: This should be a `SyntaxError`

File: Tools/cases_generator/analyzer.py [hunk 11]
```diff
@@ -1079,6 +1081,35 @@ def add_instruction(name: str) -> None:
...
+
+    If there is more than one instruction that contains the uop,
+    ensure that they all have the same size.
+    """
+    size = None
+    for inst in instructions.values():
+        if uop in inst.parts:
+            if size is None:
+                size = inst.size
+            assert size == inst.size, (
```
Comment (markshannon) [inline Tools/cases_generator/analyzer.py]: This should not be an assertion. This is an input error, not a logical one.
```suggestion
            raise analysis_error(
```

File: Tools/cases_generator/analyzer.py [hunk 12]
```diff
@@ -1079,6 +1081,35 @@ def add_instruction(name: str) -> None:
...
 
+def is_instruction_size_macro(token: lexer.Token) -> bool:
+    return token.kind == "IDENTIFIER" and token.text == "INSTRUCTION_SIZE"
+
+
+def contains_instruction_size_macro(uop: Uop) -> bool:
+    return any(is_instruction_size_macro(token) for token in uop.body)
+
+
+def get_instruction_size_for_uop(instructions: dict[str, Instruction], uop: Uop) -> int:
```
Comment (markshannon) [inline Tools/cases_generator/analyzer.py]: This should return `None` if `INSTRUCTION_SIZE` is not present.
Just move the body of `contains_instruction_size_macro` here.

File: Tools/cases_generator/analyzer.py [hunk 13]
```diff
@@ -1122,6 +1153,10 @@ def analyze_forest(forest: list[parser.AstNode]) -> Analysis:
                     continue
                 if target.text in instructions:
                     instructions[target.text].is_target = True
+    for uop in uops.values():
+        if not contains_instruction_size_macro(uop):
```
Comment (markshannon) [inline Tools/cases_generator/analyzer.py]: This shouldn't be necessary, see above comment.

**Review by markshannon — CHANGES_REQUESTED** [review]
The assertion needs to be changed to explicitly raise. Otherwise it won't work with `-O` enabled.

File: Tools/cases_generator/analyzer.py [hunk 14]
```diff
@@ -1079,6 +1081,36 @@ def add_instruction(name: str) -> None:
     return instmap, len(no_arg), min_instrumented
 
 
+def is_instruction_size_macro(token: lexer.Token) -> bool:
+    return token.kind == "IDENTIFIER" and token.text == "INSTRUCTION_SIZE"
```
Comment (markshannon) [inline Tools/cases_generator/analyzer.py]:
```suggestion
    return token.text == "INSTRUCTION_SIZE"
```
If the text is "INSTRUCTION_SIZE" then the token must be an identifier.

File: Tools/cases_generator/analyzer.py [hunk 15]
```diff
@@ -1079,6 +1081,36 @@ def add_instruction(name: str) -> None:
...
+
+
+def get_instruction_size_for_uop(instructions: dict[str, Instruction], uop: Uop) -> int | None:
+    """Return the size of the instruction that contains the given uop or
+    `None` if the uop does not contains the `INSTRUCTION_SIZE` macro.
+
+    If there is more than one instruction that contains the uop,
+    ensure that they all have the same size.
+    """
+    if not (token := next((t for t in uop.body if is_instruction_size_macro(t)), None)):
```
Comment (markshannon) [inline Tools/cases_generator/analyzer.py]:
```suggestion
    for tkn in uop.body:
          if tkn.text == "INSTRUCTION_SIZE":
               break
     else:
```

Using a generator expression, `next`, and the walrus operator makes this a bit hard to follow. Maybe use a plain for loop?

File: Tools/cases_generator/analyzer.py [hunk 16]
```diff
@@ -1079,6 +1081,36 @@ def add_instruction(name: str) -> None:
...
+        return None
+
+    size = None
+    for inst in instructions.values():
+        if uop in inst.parts:
+            if size is None:
+                size = inst.size
+            if size != inst.size:
```
Comment (markshannon) [inline Tools/cases_generator/analyzer.py:1102]:
```suggestion
            elif size != inst.size:
```

**Review by markshannon — COMMENTED** [review]
This looks good and correct now. Thanks.

I have a few suggestions, but they are mostly cosmetic.

File: Tools/cases_generator/analyzer.py (hunk 16 above)
Comment (tomasr8) [reply Tools/cases_generator/analyzer.py:1102]: I changed this to an `if` because otherwise mypy complains that the `elif` is unreachable:

```
Tools/cases_generator/analyzer.py:1103: error: Statement is unreachable  [unreachable]
```

Comment (markshannon) [reply Tools/cases_generator/analyzer.py:1102]: OK, leave it then.

**Review by markshannon — APPROVED** [review]
Looks good.
Thanks for doing this.

**Comment by tomasr8** [comment]
Thanks for all the reviews! :)
